[ALS-7694] Add DashboardDrawer feature with controller, service, and repository

TDeSain opened the pull request:
Introduce DashboardDrawerController, DashboardDrawerService, and DashboardDrawerRepository classes for dashboard drawer functionality. Include methods to fetch drawer data for datasets using materialized views.

TDeSain commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerRowMapper.java:
```diff
@@ -0,0 +1,34 @@
+
+    private List<String> convertSqlArrayToList(Array sqlArray) throws SQLException {
+        if (sqlArray == null) {
+            return List.of();
```
what

TDeSain committed:
Add Dashboard Drawer feature

- Introduce DashboardDrawerController for drawer-related endpoints
- Implement DashboardDrawerService to handle drawer logic
- Create DashboardDrawerRepository for database interactions
- Define DashboardDrawer and DashboardDrawerList records
- Add DashboardDrawerRowMapper for result set mapping
- Update DashboardRepository SQL to include dataset_id
- Modify application properties to include dashboard layout configuration

TDeSain committed:
Merge branch 'refs/heads/release' into ALS-7694

TDeSain committed:
Update application-bdc.properties with dashboard and filtering configs

- Added `dashboard.enable.bdc_hack` property
- Added `dashboard.layout.type` property
- Moved `filtering.unfilterable_concepts` property
- Ensured newline at end of file

TDeSain commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerRowMapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+public class DashboardDrawerRowMapper implements RowMapper<DashboardDrawer> {
+
```
I would at least make an abstract class for this that I could extend to fulfill multiple Dashboard Layouts.

Should probably do this for other Classes such as Dashboard.

TDeSain commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerRowMapper.java:
```diff
@@ -0,0 +1,35 @@
+        } else {
+            Object[] arrayContents = (Object[]) sqlArray.getArray();
+            // Check if the array contains a single empty value
+            if (arrayContents.length == 1 && "".equals(arrayContents[0])) {
```
This is a bit hacky but sqlArray returns an single empty value for empty values in the db.  Which makes sense as empty values can have meaning.  For this component they do not.

TDeSain committed:
Update application-bdc.properties with dashboard and filtering configs

- Added `dashboard.enable.bdc_hack` property
- Added `dashboard.layout.type` property
- Moved `filtering.unfilterable_concepts` property
- Ensured newline at end of file

TDeSain commented on src/main/resources/application-bdc.properties:
```diff
@@ -6,5 +6,8 @@ spring.datasource.username=${DATASOURCE_USERNAME}
 
 dashboard.enable.extra_details=true
+dashboard.enable.bdc_hack=true
+dashboard.layout.type=bdc
```
The dashboard enable bdc_hack was enabled in application.properties.  Should be enabled here.

Added the dashboard layout prop

Luke-Sikina commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerController.java:
```diff
@@ -0,0 +1,24 @@
+package edu.harvard.dbmi.avillach.dictionary.dashboarddrawer;
```
What's the difference between `edu.harvard.dbmi.avillach.dictionary.dashboarddrawer` and `edu.harvard.dbmi.avillach.dictionary.dashboard` as far as business purpose goes? Are these legitimately different domains, or is this a BDC specific implementation of the same domain?

TDeSain replied:
Yea I was in the same boat at the beginning and was trying to fit it into to the Dashboard thus why it was basically a replication of it until changed the class names. 

I think I should follow up on this ticket with Cleaning up some of the dashboard method and looking at how this integrates with the dashboard more closely.  

From more of a backend mindset I think it should relate to a dataset endpoint and not a feature endpoint.

Luke-Sikina commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerController.java:
```diff
@@ -0,0 +1,24 @@
+
+    @GetMapping("/{id}")
+    public ResponseEntity<DashboardDrawer> findByDatasetId(@PathVariable Integer id) {
+        return ResponseEntity.ok(dashboardDrawerService.findByDatasetId(id));
```
This should 404 if the dataset DNE. You can do this cleanly by expressing the nullable DashboardDrawer as an `Optional<DashboardDrawer>` and then mapping that to a response entity here.

Luke-Sikina commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerList.java:
```diff
@@ -0,0 +1,6 @@
+
+import java.util.List;
+
+public record DashboardDrawerList(List<DashboardDrawer> dashboardDrawerList) {
```
You made yourself a note to get rid of this record and just return a list. I agree with that note.

TDeSain replied:
I made a comment that DashboardDrawerRecord was redundant as it's already a record. 

DashboardDrawerList should be a list of DashboardDrawer as that is the collection type of this record class.    

DashboardDrawer is just a record.

Luke-Sikina commented on src/main/java/edu/harvard/dbmi/avillach/dictionary/dashboarddrawer/DashboardDrawerRepository.java:
```diff
@@ -0,0 +1,82 @@
+        try {
+            return template.query(materializedViewSql, new DashboardDrawerRowMapper());
+        } catch (Exception e) {
+            log.debug("Materialized view not available, using fallback query. Error: {}", e.getMessage());
```
Why can't we make it so this view always exists?

TDeSain replied:
https://www.postgresql.org/docs/current/rules-materializedviews.html

Materialized views are a bit different then views.  As in they are stored on disk ( in memory on some DBMS ).  They are basically a point in time view of a table.  But they need to be managed and refreshed.  

Have not use Postgres in my checkered past but they were a Saint when it came to Oracle.  Oracle has query rewrite which makes it much easier to manage.  From what I see so far in Postgres we will need to write logic blocks like this or queries will fail if Materialized view is being rewritten.

Query rewrite would fall back on the select statement used to create the materialized view if the view doesn't exist or is being refreshed.

TDeSain committed:
Add 'program_name' column to Dashboard repository

- Updated SQL query to include 'program_name' field.
- Joined 'dataset_meta' table for 'program_name' metadata.
- Added 'program_name' to the column list in DashboardService.

TDeSain committed:
Merge remote-tracking branch 'refs/remotes/origin/release' into ALS-7694

TDeSain committed:
Refactor service and controller to handle Optional returns

- Refactor `DashboardDrawerService` methods to return `Optional`
- Update controller methods to handle `Optional` results with appropriate HTTP responses
- Modify repository methods to return `Optional` for single item queries
- Ensure consistency and error handling improvements across the service and controller layers
